xapian-core/api/registry.cc

```diff
@@ -107,6 +111,49 @@ register_object(map<string, T*> & registry, const T & obj)
 }
 
 template<class T>
+static inline void
+register_object(map<string, T*> & registry1, map<string, T*> & registry2,
+		const T & obj)
+{
+    string name = obj.name();
+    if (rare(name.empty())) {
+	throw Xapian::InvalidOperationError("Unable to register object - name() method returned empty string");
+    }
+
+    string short_name = obj.name();
+    if (rare(short_name.empty())) {
+	throw Xapian::InvalidOperationError("Unable to register object - short_name() method returned empty string");
+    }
+
+    pair<typename map<string, T *>::iterator, bool> r1;
+    r1 = registry1.insert(make_pair(name, static_cast<T*>(NULL)));
+    pair<typename map<string, T *>::iterator, bool> r2;
+    r2 = registry2.insert(make_pair(short_name, static_cast<T*>(NULL)));
+
+    if (!r1.second) {
+	// Existing element with this key, so replace the pointer with NULL
+	// and delete the existing pointer.
+	//
+	// If the delete throws, this will leave a NULL entry in the map, but
+	// that won't affect behaviour as we return NULL for "not found"
+	// anyway.  The memory used will be leaked if the dtor throws, but
+	// throwing exceptions from the dtor is bad form, so that's not a big
+	// problem.
+	T * p = NULL;
+	swap(p, r1.first->second);
+	delete p;
+    }
+
+    T * clone = obj.clone();
+    if (rare(!clone)) {
+	throw Xapian::InvalidOperationError("Unable to register object - clone() method returned NULL");
+    }
+
+    r1.first->second = clone;
+    r2.first->second = clone;
+}
```

Comment:
Should we be checking for `!r1.second` instead of `r1.second` in the second condition in the check `(r1.second != r2.second || (r1.second && r1.first->second != r2.first->second))`?